### PR TITLE
Handle Material UI as plugin when shadowRoot is activated

### PR DESCRIPTION
**Pull request by nicocunin**
**What does this pull request introduce? Please describe**  
Handle Material UI styles to include them inside the shadow root of the webcomponent.

Use of JSS to do so with a StylesProvider wrapping element around the app.
-  in case useShadow is false, same behavior as before
-  in case useShadow is true but 'material-ui' plugin is not added, no wrapping, same behavior as before
-  in case useShadow is true and 'material-ui' plugin is added, wrapp app, add a mounting point to be included inside the shadowRoot, then React ShadowContent is added inside this mounting point rather than directly in the shadow root.

**How did you verify that your changes work as expected? Please describe**  
As shown afterwards with screenshots:
-  when there is no plugin "material-ui", then a raw button is displayed when using material-ui button.
-  when there is a plugin "material-ui", then a material styled button is displayed.

**Example**  
1.  Install direflkow globally
2.  Create a webcomponent with direflow: `direflow create`
3.  Follow instructions [here](https://github.com/Silind-Software/direflow/blob/master/CONTRIBUTING.md#version)
4.  Install material-ui library in the webcomponent: `yarn add @material-ui/core`
5.  Add material-ui plugin in webcomponent by modifying index.tsx. In direflowPlugins, add `{ name; "material-ui" }`
6.  Add a Material UI button in App.tsx:
```
import Button from '@material-ui/core/Button';
...
<Button variant="contained" color="primary">
   My Material UI Button
</Button>
```
7.  Serve the webcomponent `yarn start`
8.  Button appears with material-ui style and styles can be found in shadowRoot of the webcomponent

**Screenshots**  
Before:
image

After:
image

**Version**  
Based on version 3.3.0, upgraded to 3.4.0 with `yarn update-version 3.4.0`.

**NB**
Documentation will have to be updated to describe this new plugin usage.

### CODE REVIEW

**Comment by JohnDeved** [comment]
cool stuff 👍 
related to #63

**Comment by Silind** [comment]
Hi @nicocunin 
Thank you for contributing - great to see that you have come up with a solution to this issue 👍 

Would you kindly make the following changes:
- Change this PR to merge into _development_ instead of _master_.
- Make sure to merge the latest from development into your branch.
- Change version to `3.3.2` instead of `3.4.0`.

We will be back reviewing as soon as possible 😊

**Comment by nicocunin** [comment]
Hi @Silind ,

I'm happy to contribute to this cool project.

Modifications:
- Branch is now based on latest development branch instead of master
- PR is compared to development branch instead of master
- Changed version to 3.3.2

**Comment by Silind** [comment]
Great @nicocunin, looks good.
There still appears to be 4 conflicts (the 4 package.json files) - would you kindly resolve these, and push it to your branch 🙂

**Comment by nicocunin** [comment]
Conflicts resolved.

**Comment by Silind** [comment]
Perfect! 👍 We will review shortly!

We recently open-sourced the webpage as well.
If you're up for it, you're very welcome to make a PR to the documentation as well:
(https://github.com/Silind-Software/direflow-docs).

If not, it's also totally ok - we'll update it after merge 🙂

File: packages/direflow-component/src/services/materialUiStylesHandler.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import React from 'react'
+import { IDireflowPlugin } from '../types/DireflowConfig';
+
+const handleMaterialUiStyle = (app: React.FC<any> | React.ComponentClass<any, any>, plugins: IDireflowPlugin[] | undefined)
```
Comment (Silind) [inline packages/direflow-component/src/services/materialUiStylesHandler.tsx]: This line exceeds the maximum allowed characters of 120.
Please break up the parameters to align vertically.
You can use the formatter for this - ESLint should give you some heads-up 🙂

File: packages/direflow-component/src/services/materialUiStylesHandler.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import React from 'react'
```
Comment (Silind) [inline packages/direflow-component/src/services/materialUiStylesHandler.tsx]: The indentation in this file is 4 spaces. It should be 2.
You can use the formatter for this - ESLint should give you some heads-up 🙂

File: packages/direflow-component/src/services/materialUiStylesHandler.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            const mountPoint = document.createElement('span');
+
+            const jss = create({
+                ...jssPreset(),
+                insertionPoint: mountPoint
+            });
+
+            return { mountPoint, app: (<StylesProvider jss={jss}>{app}</StylesProvider>) };
+        } catch(err) {
+            console.error("Could not load Material-UI, not using them...", err);
```
Comment (Silind) [inline packages/direflow-component/src/services/materialUiStylesHandler.tsx]:
```suggestion
console.error("Could not load Material-UI", err);
```

File: packages/direflow-component/src/WebComponentFactory.tsx [hunk 4]
```diff
@@ -207,8 +208,9 @@ class WebComponentFactory {
           currentChildren = Array.from(this.children).map((child: Node) => child.cloneNode(true));
         }
 
-        const root = createProxyRoot(this);
-        ReactDOM.render(<root.open>{application}</root.open>, this);
+        const wrappedAppAndMountPoint = handleMaterialUiStyle(application, factory.plugins);
```
Comment (Silind) [inline packages/direflow-component/src/WebComponentFactory.tsx]: I'm not able to build due to typing related problems here.
`Argument of type 'Element' is not assignable to parameter of type 'FC<any> | ComponentClass<any, any>'.
  Type 'Element' is not assignable to type 'ComponentClass<any, any>'.`

File: packages/direflow-component/src/services/materialUiStylesHandler.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+            const { StylesProvider } = materialUiStyles;
+
+            const mountPoint = document.createElement('span');
+
+            const jss = create({
+                ...jssPreset(),
+                insertionPoint: mountPoint
+            });
+
+            return { mountPoint, app: (<StylesProvider jss={jss}>{app}</StylesProvider>) };
```
Comment (Silind) [inline packages/direflow-component/src/services/materialUiStylesHandler.tsx]: I'm not able to build due to typing related problems here.
`Type 'Element' is not assignable to type 'FC<any> | ComponentClass<any, any>'.
  Type 'Element' is not assignable to type 'ComponentClass<any, any>'.
    Type 'Element' provides no match for the signature 'new (props: any, context?: any): Component<any, any, any>'.`

**Comment by nicocunin** [comment]
Everything seems ok now, sorry I did not check basic ESLint issues.

**Pull request merged**